### PR TITLE
add Pow interrupt

### LINKED ISSUE #338: Add pow interrupt mechanism
**Brief of your idea**
Add pow interrupt mechanism to avoid doing something uselss.

**What could be better if XuperChain using your idea**
Decrease the possibility of forking.

**Are there any side effects?**
The mechanism lacks of some strict check.

**Additional information**
None

### PR DESCRIPTION
**Pull request by ToWorld**
## Description

What is the purpose of the change?
Add pow interrupt mechanism to avoid the frequent fork

Fixes #338 

## Type of change

- [x] Improvement

## Brief of your solution

1.Check the state regularly when minting in order to do something useless.
2.When get a new block ,check if it is valid and update the state.

## How Has This Been Tested?
Regression Test

### CODE REVIEW

File: core/valid_pow.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	if xcore == nil || block == nil || internalBlock == nil {
+		log.Warn("invalid: xcore or block or internalBlock is nil")
+		return false
+	}
+
+	// validation for consensus of pow, if ok, tell the miner to stop mining
+	newBlockHeight := internalBlock.GetHeight()
+	if xcore.con.Type(newBlockHeight) == consensus.ConsensusTypePow {
+		if newBlockHeight < xcore.Ledger.GetMeta().GetTrunkHeight() {
+			log.Warn("invalid block: new block's height is not enough")
```
Comment (fxsjy) [inline core/valid_pow.go]: it whould be better for trouble shooting, if blocks'height & trunk height are printed here.

File: core/valid_pow.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+			return false
+		}
+		actualTargetBits := internalBlock.GetTargetBits()
+		if !ledger.IsProofed(internalBlock.Blockid, actualTargetBits) {
+			log.Warn("receive a new block actual difficulty doesn't match blockid")
```
Comment (fxsjy) [inline core/valid_pow.go]: print internalBlock.Blockid and proposer here

File: ledger/calc_nonce.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+)
+
+func (l *Ledger) processFormatBlockForPOW(block *pb.InternalBlock, targetBits int32) (*pb.InternalBlock, error) {
+	var gussNonce int32
+	var gussCount int64
+	valid := false
+	var err error
+
+	for {
+		if gussCount%65535 == 0 && !l.GetPowBlockState() {
```
Comment (fxsjy) [inline ledger/calc_nonce.go]: magic number ..

File: ledger/ledger.go [hunk 4]
```diff
@@ -1128,3 +1128,24 @@ func (l *Ledger) Truncate(utxovmLastID []byte) error {
...
+
+// StartPowBlockState 更新powBlockState状态
+func (l *Ledger) StartPowBlockState() {
+	l.powMutex.Lock()
+	defer l.powMutex.Unlock()
+	l.powBlockState = true
+}
+
+// AbortPowBlockState 中断挖矿流程
+func (l *Ledger) AbortPowBlockState() {
```
Comment (fxsjy) [inline ledger/ledger.go]: maybe, AbortPowMinning is better for understanding.

File: ledger/ledger.go [hunk 5]
```diff
@@ -1128,3 +1128,24 @@ func (l *Ledger) Truncate(utxovmLastID []byte) error {
 	l.xlog.Info("truncate blockid succeed")
 	return nil
 }
+
+// StartPowBlockState 更新powBlockState状态
```
Comment (fxsjy) [inline ledger/ledger.go]: Comments for public method should be written in English.

File: ledger/ledger.go (hunk 4 above)
Comment (ToWorld) [reply ledger/ledger.go]: Yes, all suggestions above are correct and I have updated them. Thank you.

File: ledger/ledger.go (hunk 5 above)
Comment (ToWorld) [reply ledger/ledger.go]: I got it.

**Pull request merged**